### DISCUSSION TITLE
1.2.0

### DISCUSSION THREAD

**Discussion by FormerLurker** [FormerLurker/ArcWelderLib#78] [General]
At long last a new version of ArcWelderLib, ArcWelder Console, the Inverse Processor, and the associated libraries are available!

Please report any feedback in the [discussion](https://github.com/FormerLurker/ArcWelderLib/discussions) created for this release. 

# Enhancements
* Add extrusion rate variance to support Cura's Arachne engine (variable line width).  
* Add support for travel move conversion.
* Enhance inverse processor to support various firmware implementations and versions (Marlin 1/2, Prusa Firmware, Smoothieware, repetier).
* Add workaround for firmware that does not support long gcode commands.
* Reduce the max radius to reduce the impact of gcodes that are too long for the current firmware implementation.
* Enhance arc generation to convert more G1/G2 commands to arcs when transitioning from retraction/deretraction/travel.
* Make extrusion rate variance percentage (added to support cura Arachne) optional (0 = disabled).
* Enhance segment/travel statistics rendering for python code to include unicode box drawing characters.

# Arc Welder Console Argument Changes

New arguments are available for the console version of ArcWelder:

* **--allow-travel-arcs (-y)** - This will allow ArcWelder to convert travel moves to G2/G3.  Default: True
* **--extrusion-rate-variance-percent (-v)** - This will cause ArcWelder to abort any current arcs if the extrusion rate changes more than the percent (where 0.05 = 5%) specified.  This enhances quality and provides support for Cura's Arachne engine (variable line width).  0=disabled.  Default: 0.05 (5%).
* **--max-gocde-length (-c)** - Some firmware have issues with long gocde commands.  This setting allows ArcWelder to terminate arc generation if the resulting gcode contains more characters than this setting.  0=no limit.  Default = disabled.
* **--progress-type (-p)** - Supports three options:  NONE, SIMPLE, and FULL.  NONE - Prevents all progress display text.  SIMPLE - Gives just basic progress messages (percent complete), each on a single line.  FULL - Shows detailed statistics, which are useful for any third-party app that wants to give a detailed status update.  Default: SIMPLE

# Arc Straightener (Inverse Processor) Argument Changes
This has been COMPLETELY overhauled.  It's quite difficult to test these functions, so I'd appreciate any feedback you have.  

* **--firmware-type (-f)** - MARLIN_1|MARLIN_2|REPETIER|PRUSA|SMOOTHIEWARE> - Sets the firmware type to emulate when converting G2/G3 to G1/G2
* **--firmware-version (-v)** - Sets the firmware version to use.  The allowable values depend on the firmware type.  LATEST_RELEASE will use the newest version available.  Use the --help flag to see all available versions for each firmware type.  Also note that the Prusa Firmware v3.11.0 does not yet exist.  I added it in anticipation of new arc features from [this pull request](https://github.com/prusa3d/Prusa-Firmware/pull/2657) which has been added to the official milestones.  Hopefully I won't have to remove that.
* **--print-firmware-defaults (-p)** - Prints all default arc interpolation settings for the provided firmware-type and firmware-version.  This is quite useful to see what options the firmware supports.  Note that the defaults will be used unless they are overridden by other options, and you must provide a source/target to use this option until I manage to fix this limitation.  Example:  *ArcStraightener "" "" -p*:
```
Showing arguments and defaults for MARLIN_2 (LATEST_RELEASE)
Available argument for firmware: --mm-per-arc-segment, --arc-segments-per-r, --min-arc-segments, --arc-segments-per-sec, --n-arc-correction, --g90-g91-influences-extruder
Default Firmware Arguments:
        Firmware Type               : MARLIN_2
        Firmware Version            : 2.0.9.1 (LATEST_RELEASE)
        g90_g91_influences_extruder : True
        min_arc_segments            : 24
        n_arc_correction            : 25
        mm_per_arc_segment          : 1.00
        arc_segments_per_r          : 0.00
        arc_segments_per_sec        : 0.00
The following parameters do not apply to this firmware version: min_mm_per_arc_segment, mm_max_arc_error, min_circle_segments, min_arc_segment_mm, max_arc_segment_mm
```
*Note* - Not all of the following arguments are supported by all firmware implementations.   Use the --print-firmware-defaults to see if the selected firmware supports these options.
* **--max-arc-segment-mm (-d)** - Sets the maximum number of segments per millimeter.
* **--min-arc-segment-mm (-b)** - The minimum mm per arc segment.  Used to prevent unnecessarily small segments from being generated. A value less than or equal to 0 will disable this feature. Default Value: 0.
* **--min-circle-segments (-a)** - The minimum number of segments within a circle of the same radius as the arc.  Can be used to increase detail on small arcs.  The smallest segment generated will be no larger than min_mm_per_arc_segment.  A value less than or equal to 0 will disable this feature.  Default Value: 0.
* **--arc-segments-per-second (-s)** - The number of segments per second.  This will produce a constant number of arcs, clamped between mm-per-arc-segment and min-mm-per-arc-segment.  Can be used to prevent stuttering when printing very quickly.  A value less than or equal to 0 will disable this feature.  Default Value: 0.
* **--n-arc-correction (-c)** - The number of segments that will be interpolated using a small angle approximation before true sin/cos corrections are applied.  A value less than or equal to 1 will disable this feature.  Default Value: 24.
* **--min-arc-segments (-r)** - The minimum number of segments within a circle of the same radius as the arc.  Can be used to increase detail on small arcs.  The smallest segment generated will be no larger than min_mm_per_arc_segment.  A value less than or equal to 0 will disable this feature.  Default Value: 0.
* **--min-mm-per-arc-segment (-n)** - The minimum mm per arc segment.  Used to prevent unnecessarily small segments from being generated. A value less than or equal to 0 will disable this feature. Default Value: 0.
* **--mm-per-arc-segment (-m)** - The default segment length. Note that this parameter has different effects on different firmware.  Default Value: 1.00000.
* **--g90-influences-extruder (-g)** - Sets the firmware's G90/G91 influences extruder axis behavior.  By default this is determined by the firmware's behavior.  Default Value: DEFAULT.

# Important Notes about Binaries

The MacOs version may or may not work.  I've had problems building for MacOs for a long while.  If an issue is detected, and someone provides me a working build, I'll attach it.

The Raspbian version does not output the correct version info or branch, but it does include the correct build date.  This is because it is built remotely on a Raspberry Pi, which does not include git info.  I'll work to remedy this.

<hr /><em>This discussion was created from the release <a href='https://github.com/FormerLurker/ArcWelderLib/releases/tag/1.2.0'>1.2.0</a>.</em>

**Comment by fieldOfView** [comment]
On my macOS 10.12.6 (Sierra) virtual machine, the macOS binary gives the following output:
```
Aldos-Mac:~ aldo$ macOS/bin/ArcWelder --version
dyld: lazy symbol binding failed: Symbol not found: ____chkstk_darwin
  Referenced from: /Users/aldo/Downloads/macOS/bin/ArcWelder
  Expected in: /usr/lib/libSystem.B.dylib

dyld: Symbol not found: ____chkstk_darwin
  Referenced from: /Users/aldo/Downloads/macOS/bin/ArcWelder
  Expected in: /usr/lib/libSystem.B.dylib

Abort trap: 6
```
I don't know if this also happens on newer macOS versions. It would be good to hear from someone testing this on an actual mac.

Compiling the binary from source works like a charm. Here's my compiled version:
[ArcWelder_120_macOS.zip](https://github.com/FormerLurker/ArcWelderLib/files/7580038/ArcWelder_120_macOS.zip)

**Comment by FormerLurker** [comment]
Thanks @fieldOfView!  I've attached your build to the release.  At some point I'll have to figure out what's wrong with the macOs part of the github action.  Also, if you have the time, would you be willing to compile ArcStraightener (InverseProcessor)?  If so, I'll attach that to the release as well.

Also, I haven't forgotten about the API call for the ArcWelder Octoprint plugin.  I'll be working on that over the holiday.

**Comment by fieldOfView** [comment]
I had already built it, but forgot to put both in a single archive. Here's ArcStraightener:
[ArcStraightener_120_macOS.zip](https://github.com/FormerLurker/ArcWelderLib/files/7582151/ArcStraightener_120_macOS.zip)

**Reply by FormerLurker** [reply]
Oh thanks!  I'm attaching this to the release right now.

**Comment by fieldOfView** [comment]
The readme.md doesn't reflect the changes in 1.2.0 yet.

**Reply by FormerLurker** [reply]
Good catch, I'll deal with this ASAP!

**Reply by FormerLurker** [reply]
FYI, I found an error while looking through the readme.md file:  The default option for '--allow-travel-arcs' was incorrect.  It should have been disabled by default!  I will update the app ASAP, and will unfortunately need to bump the version to 1.2.1.  This only affects the console app.

**Reply by fieldOfView** [reply]
I'll make new binaries for macOS once 1.2.1 is available.

**Comment by FormerLurker** [comment]
@fieldOfView, I am thinking about switching the version to 2.0, since I made a couple potentially breaking changes.  There are some new restrictions on settings that might require some slight adjustments to your plugins.  

However, the result is a console app that's less likely to encounter any issues due to bad parameter selection.  I still need to do a bit more testing before I push.  I'll keep you posted.

**Reply by fieldOfView** [reply]
No problem, take your time.

**Comment by ETE-Design** [comment]
@FormerLurker Using Windows 11 + SuperSlicer... Before I have ran this Post-Processing: arcwelder.exe  - p -z -r=0.05 --g90-influences-extruder; but now it fails... What need to be changed to make it work again?
BTW did you ever found a way to compile the Non-Plannar Project for Windows we have talked about some time ago...

**Comment by FormerLurker** [comment]
@ETE-Design,

I realized after I published that I made a few breaking changes to some of the command line parameters (which is why I will re-release as v2.0).  The -p parameter is no longer a flag, but supports three values:

-p=NONE (this is what -p used to do by itself) - No progress updates
-p=SIMPLE - Just a simple percent complete message
-p=FULL - detailed output, which is more suitable for other programs interfacing with ArcWelder.exc

Regarding the non-plannar project, I have not yet gotten to this, but I hope to eventually.

Thanks for your comment!

**Reply by ETE-Design** [reply]
@FormerLurker Is SuperSlicer interfacing with ArcWelder, or is it only OctoPrint?

**Reply by FormerLurker** [reply]
@ETE-Design, it should work the same way as with prusa slicer.  Check the readme.  If there are differences between the SuperSlicer config and PrusaSlicer, please let me know!

**Comment by jk7gr** [comment]
Can you please release a 32bit version as well? Thank you!

**Reply by FormerLurker** [reply]
Woah, I didn't know anyone used 32bit anymore, lol!  I think I could do that, but it will take some time to add the proper action scripts.

**Comment by fieldOfView** [comment]
In december, you pushed a commit with the description "Version 2.0" (0cb60875b90c7a87b8ec9ba3b213a49b2b08ddc6), but you did not tag a release. Is version 2.0 far off? If so, I'll make an intermediate release of my Cura plugin with the 1.2 version. But I can also wait patiently.

**Reply by fieldOfView** [reply]
Is there anything I can do to help the 2.0 release along?

**Reply by FormerLurker** [reply]
I'm finally in the post-production deployment bugfix phase of this ridiculous work project now, which is going great.  Arcwelder is my next task, then the new octolapse release, finally i will have to catch up on six months of issues, lol.

Thanks for the offer.  I will let you know if I could use some help, and sorry for such an unexpectedly long delay!!!!

**Reply by ETE-Design** [reply]
@FormerLurker So how is the 2.0 Release going? Seems like the "Newest" Compiled Version in "Action" is Expired :-(
